MINOR: add column case sensitivity parameter

sushi30 opened the pull request:
- added handling for case-insensitive columns
- added handling for different numeric types (int/float/Decimal)
- added handling of boolean test case parameters

## Boolean test case parameters

Appear to be sent as json strings:

![image](https://github.com/user-attachments/assets/cb12c048-9d7b-4304-b107-eaa4760bc3b2)

Therefore I added handling as string types.

## Validation

- [x] integration tests
- [x] migrations tested on snapshot from snadbox 

<!--
Thank you for your contribution!
Unless your change is trivial, please create an issue to discuss the change before creating a PR.
-->

### Describe your changes:

Fixes <issue-number>

<!--
Short blurb explaining:
- What changes did you make?
- Why did you make them?
- How did you test your changes?
-->

I worked on ... because ...

<!-- For frontend related change, please add screenshots and/or videos of your changes preview! -->

#
### Type of change:
<!-- You should choose 1 option and delete options that aren't relevant -->
- [ ] Bug fix
- [ ] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:
<!-- add an x in [] if done, don't mark items that you didn't do !-->
- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] My PR title is `Fixes <issue-number>: <short explanation>`
- [ ] I have commented on my code, particularly in hard-to-understand areas. 
- [ ] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

<!-- Based on the type(s) of your change, uncomment the required checklist 👇 -->

<!-- Bug fix
- [ ] I have added a test that covers the exact scenario we are fixing. For complex issues, comment the issue number in the test for future reference.
-->

<!-- Improvement
- [ ] I have added tests around the new logic.
- [ ] For connector/ingestion changes: I updated the documentation.
-->

<!-- New feature
- [ ] The issue properly describes why the new feature is needed, what's the goal, and how we are building it. Any discussion
    or decision-making process is reflected in the issue.
- [ ] I have updated the documentation.
- [ ] I have added tests around the new logic.
-->

<!-- Breaking change
- [ ] I have added the tag `Backward-Incompatible-Change`.
-->

TeddyCr commented on ingestion/src/metadata/data_quality/validations/base_test_handler.py:
```diff
@@ -87,7 +88,7 @@ def get_test_case_param_value(
         )
 
-        if not value:
+        if value is None:
```
is it expected to move from considering `[]`,`""`, `None` to only considering `None`?

sushi30 replied:
retained original behavior and changed how the parameter is consumed

TeddyCr commented on ingestion/src/metadata/data_quality/validations/table/sqlalchemy/tableDiff.py:
```diff
@@ -75,6 +76,18 @@ def masked(s: str, mask: bool = True) -> str:
     return "***" if mask else s
 
 
+def is_numeric(t: type) -> bool:
+    """Check if a type is numeric.
+
+    Args:
+        t: type to check
+
+    Returns:
+        True if the type is numeric otherwise False
+    """
+    return t in [int, float, Decimal]
+
+
```
Are we able to use https://github.com/open-metadata/OpenMetadata/blob/aa7b95a515a6a72b7bad5f49aa19963b85c25be7/ingestion/src/metadata/profiler/orm/registry.py#L159 ?

sushi30 replied:
no. that is specific for sqlalchemy types

TeddyCr commented on ingestion/src/metadata/utils/collections.py:
```diff
@@ -0,0 +1,27 @@
+    """
+
+    def __eq__(self, other):
+        return self.lower() == other.lower()
+
+    def __hash__(self):
+        return hash(self.lower())
```
You should use `.casefold()` for caseless matching. https://docs.python.org/3/library/stdtypes.html#str.casefold.

e.g.
```python
>>> "fooßBar".lower()
'fooßbar'
>>> "fooßBar".casefold()
'foossbar'
```

sushi30 commented:
@TeddyCr addressed your comments:

- use casefold()
- reverted change to test case parameters
- added special handling for boolean test case parameters

TeddyCr commented on ingestion/src/metadata/data_quality/validations/utils.py:
```diff
@@ -0,0 +1,57 @@
+
+
+def get_test_case_param_value(
+    test_case_param_vals: list[TestCaseParameterValue],
```
I think you may need to make that `List[...]` vs `list[...]`

TeddyCr commented on ingestion/src/metadata/data_quality/validations/base_test_handler.py:
```diff
@@ -65,37 +66,18 @@ def run_validation(self) -> TestCaseResult:
         default: Optional[R] = None,
         pre_processor: Optional[Callable] = None,
     ) -> Optional[Union[R, T]]:
-        """Give a column and a type return the value with the appropriate type casting for the
```
Should we keep this doc string?

sushi30 replied:
I moved it to the underlying utils library bc this doesnt have much meat. didnt want to keep redundant material. wdyt?

sushi30 replied:
here: https://github.com/open-metadata/OpenMetadata/pull/18115/files#diff-81b87512f0e5a37568c837e4c3631ae91431312ff2b185e864aa18a260d1c8faR20
